bgpd/bgp_vty.c

```diff
@@ -7869,17 +7869,153 @@ static void bgp_show_bestpath_json(struct bgp *bgp, json_object *json)
 	json_object_object_add(json, "bestPath", bestpath);
 }
 
+/* Print the error code/subcode for why the peer is down */
+static void bgp_show_peer_reset(struct vty * vty, struct peer *peer,
+				json_object *json_peer, bool use_json)
+{
+	const char *code_str;
+	const char *subcode_str;
+
+	if (use_json) {
+		if (peer->last_reset == PEER_DOWN_NOTIFY_SEND
+		    || peer->last_reset == PEER_DOWN_NOTIFY_RECEIVED) {
+			char errorcodesubcode_hexstr[5];
+			char errorcodesubcode_str[256];
+
+			code_str = bgp_notify_code_str(peer->notify.code);
+			subcode_str = bgp_notify_subcode_str(
+					 peer->notify.code,
+					 peer->notify.subcode);
+
+			sprintf(errorcodesubcode_hexstr, "%02X%02X",
+				peer->notify.code, peer->notify.subcode);
+			json_object_string_add(json_peer,
+					       "lastErrorCodeSubcode",
+					       errorcodesubcode_hexstr);
+			snprintf(errorcodesubcode_str, 255, "%s%s",
+				 code_str, subcode_str);
+			json_object_string_add(json_peer,
+					       "lastNotificationReason",
+					       errorcodesubcode_str);
+			if (peer->last_reset == PEER_DOWN_NOTIFY_RECEIVED
+			    && peer->notify.code == BGP_NOTIFY_CEASE
+			    && (peer->notify.subcode
+				== BGP_NOTIFY_CEASE_ADMIN_SHUTDOWN
+				|| peer->notify.subcode
+				== BGP_NOTIFY_CEASE_ADMIN_RESET)
+			    && peer->notify.length) {
+				char msgbuf[1024];
+				const char *msg_str;
+
+				msg_str = bgp_notify_admin_message(
+					     msgbuf, sizeof(msgbuf),
+					     (uint8_t *)peer->notify.data,
+					     peer->notify.length);
+				if (msg_str)
+					json_object_string_add(
+					   json_peer,
+					   "lastShutdownDescription",
+					   msg_str);
+			}
+
+		} 
+		json_object_string_add(json_peer, "lastResetDueTo",
+				       peer_down_str[(int)peer->last_reset]);
+	} else {
+		if (peer->last_reset == PEER_DOWN_NOTIFY_SEND
+		    || peer->last_reset == PEER_DOWN_NOTIFY_RECEIVED) {
+			code_str = bgp_notify_code_str(peer->notify.code);
+			subcode_str =
+				bgp_notify_subcode_str(peer->notify.code,
+						       peer->notify.subcode);
+			vty_out(vty, "  Notification %s (%s%s)\n",
+				peer->last_reset == PEER_DOWN_NOTIFY_SEND
+				? "sent"
+				: "received",
+				code_str, subcode_str);
+		} else {
+			vty_out(vty, "  %s\n",
+				peer_down_str[(int)peer->last_reset]);
+		}
+	}
+}
+
+static inline bool bgp_has_peer_failed(struct peer *peer, afi_t afi,
+				       safi_t safi)
+{
+	return ((peer->status != Established) ||
+		!peer->afc_recv[afi][safi]);
+}
+
+static void bgp_show_failed_summary(struct vty *vty, struct bgp *bgp,
+				    struct peer *peer, json_object *json_peer,
+				    int max_neighbor_width, bool use_json)
+{
+	char timebuf[BGP_UPTIME_LEN], dn_flag[2];
+	int len;
+
+	if (use_json) {
+		if (peer_dynamic_neighbor(peer)) {
```

Comment:
I see {} not used below for a single statement (just a minor notice)

Reply:
Fixed